### PR TITLE
Fix listing tags bug

### PR DESCRIPTION
**Pull request by sai-bhargav**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
This PR will fix the listing tags bug. Changes made in the query for fetching data from ES.After the Change only if all the tags are present the article will be fetched.

## Related Tickets & Documents

Closes https://github.com/thepracticaldev/dev.to/issues/7513

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

**Comment by rhymes** [comment]
Hi @sai-bhargav, are you referring to https://github.com/thepracticaldev/dev.to/issues/7513 ?

**Comment by sai-bhargav** [comment]
> Hi @sai-bhargav, are you referring to #7513 ?

yes
https://github.com/thepracticaldev/dev.to/issues/7513

File: app/services/search/query_builders/classified_listing.rb [hunk 1]
```diff
@@ -78,6 +84,10 @@ def range_keys
       def range_keys_present?
         RANGE_KEYS.detect { |key| @params[key].present? }
       end
+
+      def tags_key?(term_key)
+        (term_key == TERM_KEYS[5]) && (@params[term_key].is_a? Array)
```
Comment (mstruve) [inline app/services/search/query_builders/classified_listing.rb:89]: In the future, I would love the ability to be able to switch between an AND query and an OR query. With that in mind, I think a better solution here might be to support both. [Here is how I handled fixing this issue for the ReadingList tag filters.](https://github.com/thepracticaldev/dev.to/pull/7425)
```
      def terms_keys
  TERM_KEYS.flat_map do |term_key, search_key|
    next unless @params.key? term_key

    values = Array.wrap(@params[term_key])

    if params[:tag_boolean_mode] == "all" && term_key == :tag_names
      values.map { |val| { term: { search_key => val } } }
    else
      { terms: { search_key => values } }
    end
  end.compact
end
```

Comment (sai-bhargav) [reply app/services/search/query_builders/classified_listing.rb]: hey @mstruve , thanks for the suggestion. Refactored the code accordingly

Comment (mstruve) [reply app/services/search/query_builders/classified_listing.rb]: Excellent! Its getting there but we are still missing the param `tag_boolean_mode` which will tell the code whether we want to set up the filter for an AND vs OR Elasticsearch query.

File: spec/services/search/query_builders/classified_listing_spec.rb [hunk 2]
```diff
@@ -16,11 +16,25 @@
 
   describe "#as_hash" do
     it "applies TERM_KEYS from params" do
-      params = { category: "cfp", tags: ["beginner"], contact_via_connect: false }
+      params = { category: "cfp", tags: %w[beginner intermediate professional], contact_via_connect: false }
+      filter = described_class.new(params: params)
+      exepcted_filters = [
```
Comment (rhymes) [inline spec/services/search/query_builders/classified_listing_spec.rb:21]: small typo here `exepcted_filters`

File: spec/services/search/query_builders/classified_listing_spec.rb [hunk 3]
```diff
@@ -16,11 +16,25 @@
...
+        { "terms" => { "contact_via_connect" => [false] } },
+        { "terms" => { "published" => [true] } },
+      ]
+      expect(filter.as_hash.dig("query", "bool", "filter")).to match_array(exepcted_filters)
+    end
+
+    it "applies TERM_KEYS from params with boolean mode" do
+      params = { category: "cfp", tags: %w[beginner intermediate professional], contact_via_connect: false, tag_boolean_mode: "all" }
       filter = described_class.new(params: params)
       exepcted_filters = [
```
Comment (rhymes) [inline spec/services/search/query_builders/classified_listing_spec.rb:33]: small typo here `exepcted_filters`

File: spec/services/search/query_builders/classified_listing_spec.rb (hunk 2 above)
Comment (sai-bhargav) [reply spec/services/search/query_builders/classified_listing_spec.rb:21]: It's spelled the same at various parts in the codebase. Thought it was some kind of convention being followed and intentionally spelled like that.

some places i have found apart from this spec file

spec/services/search/query_builders/feed_content_spec.rb
spec/services/search/query_builders/reaction_spec.rb
spec/services/search/query_builders/user_spec.rb

will create a new PR to fix the typos if required.

File: spec/services/search/query_builders/classified_listing_spec.rb (hunk 3 above)
Comment (sai-bhargav) [reply spec/services/search/query_builders/classified_listing_spec.rb:33]: It's spelled the same at various parts in the codebase. Thought it was some kind of convention being followed and intentionally spelled like that.

some places i have found apart from this spec file

spec/services/search/query_builders/feed_content_spec.rb
spec/services/search/query_builders/reaction_spec.rb
spec/services/search/query_builders/user_spec.rb

will create a new PR to fix the typos if required.

Comment (rhymes) [reply spec/services/search/query_builders/classified_listing_spec.rb:33]: thank you1

**Pull request merged**